src/wallet/wallet.cpp

```diff
@@ -784,6 +784,27 @@ bool CWallet::IsSpentKey(const uint256& hash, unsigned int n) const
     return false;
 }
 
+void CWallet::SetUsedAddressState(WalletBatch& batch, const CTxOut& cout, bool used)
+{
+    AssertLockHeld(cs_wallet);
+    CTxDestination dst;
+    if (ExtractDestination(cout.scriptPubKey, dst)) {
+        if (IsMine(dst)) {
+            if (used && !GetDestData(dst, "address_used", nullptr)) {
+                AddDestData(batch, dst, "address_used", "p"); // p for "present", opposite of absent (null)
+            } else if (!used && GetDestData(dst, "address_used", nullptr)) {
+                EraseDestData(batch, dst, "address_used");
+            }
+        }
+    }
+}
+
+bool CWallet::IsUsedAddress(const CTxDestination& dst) const
+{
+    LOCK(cs_wallet);
```

Comment:
Is this lock needed?

Reply:
Probably not needed as `GetDestData` is just accessing the address book data. I have now removed the lock.

Reply:
Sorry about that; it appears to be needed.
```
wallet/wallet.cpp:803:12: error: calling function 'GetDestData' requires holding mutex 'cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
    return GetDestData(dst, "address_used", nullptr);
           ^
1 error generated.
```

Reply:
My bad, should have checked properly myself before removing! This has now been updated to include the lock again.